docs: cicd example guide (#903)
langchain-ai/docs: src/langsmith/cicd-pipeline-example.mdx

```mermaid
graph TD
    A1[Code or Graph Change] --> B1[Trigger CI Pipeline]
    A2[Prompt Commit in PromptHub] --> B1
    A3[Online Evaluation Alert] --> B1
    A4[PR Opened] --> B1

    subgraph "Testing"
        B1 --> C1[Run Unit Tests]
        B1 --> C2[Run Integration Tests]
        B1 --> C3[Run End to End Tests]
        B1 --> C4[Run Offline Evaluations]

        C4 --> D1[Evaluate with OpenEvals or AgentEvals]
        C4 --> D2[Assertions: Hard and Soft]

        C1 --> E1[Run LangGraph Dev Server Test]
        C2 --> E1
        C3 --> E1
        D1 --> E1
        D2 --> E1
    end

    E1 --> F1[Push to Staging Deployment - Deploy to LangSmith as Development Type]

    F1 --> G1[Run Online Evaluations on Live Data]
    G1 --> H1[Attach Scores to Traces]

    H1 --> I1[If Quality Below Threshold]
    I1 --> J1[Send to Annotation Queue]
    I1 --> J2[Trigger Alert via Webhook]
    I1 --> J3[Push Trace to Golden Dataset]

    F1 --> K1[Promote to Production if All Pass - Deploy to LangSmith Production]

    J2 --> L1[Slack or PagerDuty Notification]

    subgraph Manual Review
        J1 --> M1[Human Labeling]
        M1 --> J3
    end
```

> ### Trigger Sources

Reviewer: ```suggestion
### Trigger sources
```

There are multiple ways you can trigger this pipeline, either during development or if your application is already live. The pipeline can be triggered by:

- <Icon icon="code-branch" /> **Code Changes**: Pushes to main/develop branches where you can modify the LangGraph architecture, try different models, update agent logic, or make any code improvements
- <Icon icon="edit" /> **PromptHub Updates**: Changes to prompt templates stored in LangSmith PromptHub - whenever there's a new prompt commit, we trigger a webhook to run the pipeline
- <Icon icon="exclamation-triangle" /> **Online Evaluation Alerts**: Performance degradation notifications from live deployments
- <Icon icon="webhook" /> **LangSmith Traces Webhooks**: Automated triggers based on trace analysis and performance metrics

> - <Icon icon="play" /> **Manual Trigger**: Manual initiation of the pipeline for testing or emergency deployments

Reviewer: ```suggestion
- <Icon icon="code-branch" /> **Code changes**: Pushes to main/development branches where you can modify the LangGraph architecture, try different models, update agent logic, or make any code improvements.
- <Icon icon="edit" /> **PromptHub updates**: Changes to prompt templates stored in LangSmith PromptHub—whenever there's a new prompt commit, we trigger a webhook to run the pipeline.
- <Icon icon="exclamation-triangle" /> **Online evaluation alerts**: Performance degradation notifications from live deployments
- <Icon icon="webhook" /> **LangSmith traces webhooks**: Automated triggers based on trace analysis and performance metrics.
- <Icon icon="play" /> **Manual trigger**: Manual initiation of the pipeline for testing or emergency deployments.
```

### Testing Layers

Compared to more traditional software, testing AI agent applications includes assessing the quality of the agent responses and therefore it is essential to evaluate each part of the workflow. The pipeline implements multiple testing layers:

1. <Icon icon="puzzle-piece" /> **Unit Tests**: Individual node and utility function testing
2. <Icon icon="link" /> **Integration Tests**: Component interaction testing
3. <Icon icon="route" /> **End-to-End Tests**: Full graph execution testing
4. <Icon icon="brain" /> **Offline Evaluations**: Performance assessment with real-world scenarios including end-to-end evaluations, single-step evaluations, agent trajectory analysis, and multi-turn simulations

> 5. <Icon icon="server" /> **LangGraph Dev Server Tests**: We use our [langgraph-cli](/langsmith/cli) tool for spinning up (inside the GitHub Action) a local server to run the LangGraph agent. We poll the `/ok` server API endpoint until it is available and for 30 seconds, after that we throw an error

Reviewer 2: Yes I'm good with removing the We

## Prerequisites

Before setting up the CI/CD pipeline, ensure you have:

- <Icon icon="robot" /> An AI agent application (in this case built using [LangGraph](/oss/langgraph/overview))
- <Icon icon="key" /> [LangSmith account](/langsmith/create-account-api-key) and API key needed to deploy agents and retrieve experiment results
- <Icon icon="cog" /> Additional environment variables configured in your repository secrets

<Note>
While this example uses GitHub, the CI/CD pipeline works with other Git hosting platforms including GitLab, Bitbucket, and others.
</Note>

## GitHub Actions Workflow

The CI/CD pipeline leverages GitHub Actions to automate the deployment process, utilizing the [Control Plane API](/langsmith/api-ref-control-plane) and [LangSmith API](/langsmith/api-reference) for seamless integration. The implementation includes a comprehensive helper script available at [https://github.com/langchain-ai/cicd-pipeline-example/blob/main/.github/scripts/langgraph_api.py](https://github.com/langchain-ai/cicd-pipeline-example/blob/main/.github/scripts/langgraph_api.py) that facilitates API interactions and deployment management.

> ### New Agent Deployment

Reviewer: Could link to offline evaluations and dev server too:
/langsmith/evaluation-concepts#offline-evaluation
/langsmith/local-server

    Analyzes the complete path your agent takes through the graph, including all intermediate steps and decision points. This helps identify bottlenecks, unnecessary steps, or suboptimal routing in your agent's workflow. It also evaluates whether your agent invoked the right tools in the right order or at the right time.
  </Accordion>

  <Accordion title="Multi-Turn Evaluation" icon="comments">
    Tests conversational flows where the agent maintains context across multiple interactions. This is crucial for agents that handle follow-up questions, clarifications, or extended dialogues with users.
  </Accordion>
</AccordionGroup>

See the [LangGraph testing documentation](/oss/python/langgraph/test) for specific testing approaches and the [evaluation approaches guide](/langsmith/evaluation-approaches) for a comprehensive overview of offline evaluations.

## Deployment Options

We support multiple deployment methods, mostly based on the LangSmith instance you have:

- <Icon icon="cloud" /> **Cloud LangSmith**: Direct GitHub integration or Docker image deployment
- <Icon icon="server" /> **Self-Hosted/Hybrid**: Container registry-based deployments

> The deployment flow starts by modifying your agent implementation. At minimum, you must have a `langgraph.json` and dependency file in your project (`requirements.txt` or `pyproject.toml`). We use the `langgraph dev` CLI tool to check for errors - if there are any, we fix the issues; otherwise, the deployment will likely succeed when deployed to LangSmith Deployments.

Reviewer: ```suggestion
The deployment flow starts by modifying your agent implementation. At minimum, you must have a [`langgraph.json`](/langsmith/application-structure) and dependency file in your project (`requirements.txt` or `pyproject.toml`). Use the `langgraph dev` CLI tool to check for errors—fix any errors; otherwise, the deployment will succeed when deployed to LangSmith Deployments.
```

    B --> C[Test Locally with langgraph dev]
    C --> D{Errors?}
    D -->|Yes| E[Fix Issues]
    E --> C
    D -->|No| F[Choose LangSmith Instance]

    F --> G[Cloud LangSmith]
    F --> H[Self-Hosted/Hybrid LangSmith]

    subgraph "Cloud LangSmith"
        G --> I[Method 1: Connect GitHub Repo in UI]
        G --> J[Method 2: Docker Image]
        I --> K[Deploy via LangSmith UI]
        J --> L[Build Docker Image langgraph build]
        L --> M[Push to Container Registry]
        M --> N[Deploy via Control Plane API]
    end

    subgraph "Self-Hosted/Hybrid LangSmith"
        H --> S[Build Docker Image langgraph build]
        S --> T[Push to Container Registry]
        T --> U{Deploy via?}
        U -->|UI| V[Specify Image URI in UI]
        U -->|API| W[Use Control Plane API]
        V --> X[Deploy via LangSmith UI]
        W --> Y[Deploy via Control Plane API]
    end

    K --> AA[Agent Ready for Use]
    N --> AA
    X --> AA
    Y --> AA

    AA --> BB{Connect via?}
    BB -->|LangGraph SDK| CC[Use LangGraph SDK]
    BB -->|RemoteGraph| DD[Use RemoteGraph]
    BB -->|REST API| EE[Use REST API]
    BB -->|LangGraph Studio UI| FF[Use LangGraph Studio UI]
```

> ### Prerequisites for Manual Deployment

Reviewer: ```suggestion
### Prerequisites for manual deployment
```

    "dependencies": ["."],
    "image_distro": "wolfi"
}
```

### Method 1: LangSmith Deployment UI (Cloud Only)

Deploy your agent using the LangSmith deployment interface for cloud deployments:

1. Go to your LangSmith dashboard
2. Navigate to the Deployments section
3. Connect your GitHub repository and specify the agent path

<Info>
**Benefits:**
- Simple UI-based deployment
- Direct integration with your GitHub repository
- No manual Docker image management required
</Info>

### Method 2: Build Docker Image with LangGraph CLI

Build a Docker image directly using the LangGraph CLI:

```bash
# Build Docker image
langgraph build -t my-agent:latest

# Push to your container registry
docker push my-agent:latest
```

You can push to any container registry (Docker Hub, AWS ECR, Azure ACR, Google GCR, etc.) that your deployment environment has access to.

**Deployment Options:**
- **Cloud LangSmith**: Use the Control Plane API to create deployments from your container registry
- **Self-Hosted/Hybrid LangSmith**: Choose between LangSmith UI or Control Plane API

See the [LangGraph CLI build documentation](/langsmith/cli#build) for more details.

### Local Development & Testing

> ![LangGraph Studio CLI Interface](./images/cicd-studio-cli.png)

Reviewer: If it's possible to use a screenshot that is updated without "LangGraph" in the Studio name, that would be awesome. Otherwise, we can update it later.
